pkg/wait/poll_watcher.go

```diff
@@ -0,0 +1,189 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wait
+
+import (
+	"sync"
+	"time"
+
+	api_errors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+)
+
+// PollInterval determins when you should poll.  Useful to mock out, or for
+// replacing with exponential backoff later.
+type PollInterval interface {
+	PollChan() <-chan time.Time
+	Stop()
+}
+
+type pollingWatcher struct {
+	c        rest.Interface
+	ns       string
+	resource string
+	name     string
+	timeout  time.Duration
+	done     chan bool
+	result   chan watch.Event
+	wg       *sync.WaitGroup
+	// we can mock the interface for testing.
+	pollInterval PollInterval
+	// mock hook for testing.
+	poll func() (runtime.Object, error)
+}
+
+type watchF func(v1.ListOptions) (watch.Interface, error)
+
+type tickerPollInterval struct {
+	t *time.Ticker
+}
+
+func (t *tickerPollInterval) PollChan() <-chan time.Time {
+	return t.t.C
+}
+
+func (t *tickerPollInterval) Stop() {
+	t.t.Stop()
+}
+
+func newTickerPollInterval(d time.Duration) *tickerPollInterval {
+	return &tickerPollInterval{time.NewTicker(d)}
+}
+
+// NewWatcher makes a watch.Interface on the given resource in the client,
+// falling back to polling if the server does not support Watch.
+func NewWatcher(watchFunc watchF, c rest.Interface, ns string, resource string, name string, timeout time.Duration) (watch.Interface, error) {
+	native, err := nativeWatch(watchFunc, name, timeout)
+	if err == nil {
+		return native, nil
+	}
+	polling := &pollingWatcher{
+		c, ns, resource, name, timeout, make(chan bool), make(chan watch.Event), &sync.WaitGroup{},
+		newTickerPollInterval(time.Second), nativePoll(c, ns, resource, name)}
+	err = polling.start()
+	if err != nil {
+		return nil, err
+	}
+	return polling, nil
+}
+
+func (w *pollingWatcher) start() error {
+	w.wg.Add(1)
+
+	go func() {
+		defer w.wg.Done()
+		defer w.pollInterval.Stop()
+		var err error
+		var old, new runtime.Object
+		done := false
+		for !done {
+			old = new
+
+			select {
+			case <-w.pollInterval.PollChan():
+				new, err = w.poll()
+				newObj, ok1 := new.(v1.Object)
+				oldObj, ok2 := old.(v1.Object)
+
+				if err != nil && api_errors.IsNotFound(err) {
+					if old != nil {
+						// Deleted
+						w.result <- watch.Event{
+							Type:   watch.Deleted,
+							Object: old,
+						}
+					}
+					//... Otherwise maybe just doesn't exist.
+				} else if err != nil {
+					// Just an error
+					w.result <- watch.Event{
+						Type: watch.Error,
+					}
+				} else if old == nil && new != nil {
+					// Added
+					w.result <- watch.Event{
+						Type:   watch.Added,
+						Object: new,
+					}
+				} else if !(ok1 && ok2) {
```

Comment:
Should we separate the if check for `err`, `ok1` and `ok2` ? (OR may be a nested switch)